### PR TITLE
Dpnp ptp

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

**Review by shssf — APPROVED** [review]

**Review by densmirn — APPROVED** [review]

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 1]
```diff
@@ -163,6 +163,39 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+        return;
+    }
+
+    if (ndim < 1)
+    {
+        return;
+    }
+
+    _DataType* min_arr = reinterpret_cast<_DataType*>(dpnp_memory_alloc_c(size * sizeof(_DataType)));
+    _DataType* max_arr = reinterpret_cast<_DataType*>(dpnp_memory_alloc_c(size * sizeof(_DataType)));
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:191]: Where is freeing of this memory?

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 2]
```diff
@@ -163,6 +163,39 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+
+    if ((arr == nullptr) || (result == nullptr))
+    {
+        return;
+    }
+
+    if (ndim < 1)
+    {
+        return;
+    }
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:188]: Checks like these, that are forward to immediately return from function, should be in the begin of function before any other computations.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 3]
```diff
@@ -163,6 +163,39 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+    {
+        size *= shape[i];
+    }
+
+    DPNPC_ptr_adapter<_DataType> input1_ptr(array1_in, size, true);
+    DPNPC_ptr_adapter<_DataType> result_ptr(result1, size, true, true);
+    _DataType* arr = input1_ptr.get_ptr();
+    _DataType* result = result_ptr.get_ptr();
+
+    if ((arr == nullptr) || (result == nullptr))
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:180]: It looks like that `arr == nullptr` only when `array1_in == nullptr`, is not it?

**Review by Alexander-Makaryev — CHANGES_REQUESTED** [review]

File: dpnp/dpnp_iface_arraycreation.py [hunk 4]
```diff
@@ -1082,6 +1083,37 @@ def ones_like(x1, dtype=None, order='C', subok=False, shape=None):
...
+        elif out is not None:
+            pass
+        elif keepdims is not numpy._NoValue:
+            pass
+        else:
+            result = dpnp_ptp(arr_desc, axis=axis).get_pyobj()
+   
+            # scalar returned
+            if result.shape == (1,):
+                return result.dtype.type(result[0])
```
Comment (densmirn) [inline dpnp/dpnp_iface_arraycreation.py:1110]: You could call function `dpnp.convert_single_elem_array_to_scalar` here.

File: dpnp/dpnp_iface_arraycreation.py [hunk 5]
```diff
@@ -1082,6 +1083,37 @@ def ones_like(x1, dtype=None, order='C', subok=False, shape=None):
...
+
+    Limitations
+    -----------
+    Input array is supported as :obj:`dpnp.ndarray`.
+    Parameters ``out`` and ``keepdims`` are supported only with default values.
+    """
+    arr_desc = dpnp.get_dpnp_descriptor(arr)
+    if arr_desc:
+        if axis is not None and not isinstance(axis, int):
+            pass
```
Comment (densmirn) [inline dpnp/dpnp_iface_arraycreation.py:1100]: You could decrease indentations with the following way:
```suggestion
    if not arr_desc:
        pass
    elif axis is not None and not isinstance(axis, int):
        pass
```

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 6]
```diff
@@ -163,6 +163,42 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+    }
+
+    size_t size = 1;
+    for (size_t i = 0; i < ndim; ++i)
+    {
+        size *= shape[i];
+    }
+
+    DPNPC_ptr_adapter<_DataType> input1_ptr(array1_in, size, true);
+    DPNPC_ptr_adapter<_DataType> result_ptr(result1, size, true, true);
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:186]: Do we really need `true` parameters for these adapters?

File: dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx [hunk 7]
```diff
@@ -272,6 +272,59 @@ cpdef utils.dpnp_descriptor dpnp_ones_like(result_shape, result_dtype):
...
+    if axis_ is not None:
+        axis1 = axis_
+        axis2.reserve(len(axis1))
+        for shape_it in axis1:
+            if shape_it < 0:
+                raise ValueError("DPNP dparray::__init__(): Negative values in 'shape' are not allowed")
+            axis2.push_back(shape_it)
+        axis_size = len(axis1)
+    
+    func(result.get_data(), arr.get_data(), < size_t * > output_shape.data(), result.ndim, < size_t * > shape_arr.data(), arr.ndim, < size_t * > axis2.data(), axis_size)
```
Comment (densmirn) [inline dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx:323]: Do you really need casting to `< size_t * >`?

File: dpnp/backend/include/dpnp_iface.hpp [hunk 8]
```diff
@@ -457,6 +457,22 @@ INP_DLLEXPORT void dpnp_prod_c(void* result_out,
...
+ * @param [in]  array       Input array with data.
+ * @param [in]  res_shape   Shape of input array.
+ * @param [in]  res_ndim    Number of elements in shape.
+ * @param [in]  shape       Shape of input array.
+ * @param [in]  ndim        Number of elements in shape.
+ * @param [in]  axis        Axis.
+ * @param [in]  naxis       Number of elements in axis.
+ */
+template <typename _DataType>
+void dpnp_ptp_c(void* result1, void* array1_in, const size_t* res_shape, size_t res_ndim, const size_t* shape, size_t ndim, const size_t* axis, size_t naxis);
```
Comment (densmirn) [inline dpnp/backend/include/dpnp_iface.hpp:474]: Let's make interface of the function similar to interface of `dpnp_matmul_c`. I mean result and input parameters. Parameters `axis` look ok.

**Comment by Alexander-Makaryev** [comment]
@LukichevaPolina Could you please mark conversations above that were processed? It is hard to understand what is done and what is not. I set "In progress" label, please remove it and request review one more time when you will be ready.

File: dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx (hunk 7 above)
Comment (densmirn) [reply dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx:323]: Looks like problem is in `custom_arraycreation_1in_1out_func_ptr_t`. Could you please change types `size_t *` to `long *` there?

**Review by Alexander-Makaryev — APPROVED** [review]
LGTM

**Pull request merged**